### PR TITLE
Multiple url parameters

### PR DESCRIPTION
**Pull request by chunyu3**
# Description
Fix https://github.com/Azure/autorest.csharp/issues/2730

- support url which contains more than one url parameter
- 
we need to support following Url schemas:
1. http://{urlstr}
2. http://{urlstr}/parameter/{parameterName}
3. http://{namespace}.service.com
4. http://{namespace}.service.com/parameter/{parameterName}

For #1, #2, we still change `urlStr` to `endpoint` with Url type and isEndpoint=true. And `parameterName` is a normal client parameter with String type, isEndpoint=false
For #3, #4, both `namespace` and `parameterName` are normal client parameter with string type, and isEndpoint=false

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

**Comment by mpodwysocki** [comment]
@chunyu3 When trying to compile the C# CADL using this repo for [Azure Notification Hubs](https://github.com/mpodwysocki/azure-notificationhubs-cadl), I get the following:

```bash
> azure-notificationhubs-cadl@1.0.0 build:csharp
> cadl compile main.cadl --emit ../autorest.csharp.fork/src/CADL.Extension/Emitter.Csharp/dist/src/index.js --output-path cadl-output/csharp

Cadl compiler v0.35.0

No API-Version provided.
Compilation completed successfully, output files are in ~/git/azure-notificationhubs-cadl/cadl-output/csharp.
```

This seems new where before it was picking up the API version successfully.  See the current [`cadl.json`](https://github.com/mpodwysocki/azure-notificationhubs-cadl/blob/main/cadl-output/csharp/cadl.json) where it picks up the API version successfully.

**Comment by chunyu3** [comment]
> ```shell
>  main.cadl --emit ../autorest.csharp.fork/src/CADL.Extension/Emitter.Csharp/dist/src/index.js --output-path cadl-output/csharp
> ```

I tried locally, but I cannot reproduce this issue. @mpodwysocki would you please try again? if it still has problem, we can setup meeting and work together for this, thanks.

**Comment by mpodwysocki** [comment]
@chunyu3 Ok, rechecked and all works as expected.  Thanks

**Comment by chunyu3** [comment]
> @chunyu3 When trying to compile the C# CADL using this repo for [Azure Notification Hubs](https://github.com/mpodwysocki/azure-notificationhubs-cadl), I get the following:
> 
> ```shell
> > azure-notificationhubs-cadl@1.0.0 build:csharp
> > cadl compile main.cadl --emit ../autorest.csharp.fork/src/CADL.Extension/Emitter.Csharp/dist/src/index.js --output-path cadl-output/csharp
> 
> Cadl compiler v0.35.0
> 
> No API-Version provided.
> Compilation completed successfully, output files are in ~/git/azure-notificationhubs-cadl/cadl-output/csharp.
> ```
> 
> This seems new where before it was picking up the API version successfully. See the current [`cadl.json`](https://github.com/mpodwysocki/azure-notificationhubs-cadl/blob/main/cadl-output/csharp/cadl.json) where it picks up the API version successfully.

confirmed. This issue does not occur. close it.

**Comment by pshao25** [comment]
What is the difference between #1 and #3?

In #2, is `parameterName` a client parameter?

**Comment by chunyu3** [comment]
> What is the difference between #1 and #3?
> 
> In #2, is `parameterName` a client parameter?

#1 the parameter urlstr is the whole endpoint, #3, the parameter is just part of the endpoint.
in all scenarios,  the parameterName is parameter of url, it is client parameter.

File: src/CADL.Extension/Emitter.Csharp/src/lib/cadlServer.ts [hunk 1]
```diff
@@ -38,18 +38,19 @@ export function resolveServers(
     return servers.map((server) => {
         const parameters: InputParameter[] = [];
         let url: string = server.url;
+        let endpoint: string = url.replace("http://", "").replace("https://", "").split("/")[0];
         for (const [name, prop] of server.parameters) {
             // if (!validateValidServerVariable(program, prop)) {
             //   continue;
             // }
-
+            const isEndpoint: boolean = endpoint === (`{${name}}`);
```
Comment (AlexanderSher) [inline src/CADL.Extension/Emitter.Csharp/src/lib/cadlServer.ts:46]: We probably can be even more specific and check for name `"$host"`, similar to M4: https://github.com/Azure/autorest/blob/5149dd0311cd5fad437ac18cff1e387b0835c1a5/packages/extensions/modelerfour/src/modeler/modelerfour.ts#L1748

Comment (chunyu3) [reply src/CADL.Extension/Emitter.Csharp/src/lib/cadlServer.ts:46]: > We probably can be even more specific and check for name `"$host"`, similar to M4: https://github.com/Azure/autorest/blob/5149dd0311cd5fad437ac18cff1e387b0835c1a5/packages/extensions/modelerfour/src/modeler/modelerfour.ts#L1748

Hello @AlexanderSher cadl will not define host parameter as global client parameter, it only define `@server` to define url. So we may not need to check this.

**Review by AlexanderSher — APPROVED** [review]

**Pull request merged**